internal/provider/data_logging_endpoints.go

```diff
@@ -0,0 +1,193 @@
+		)
+		return
+	}
+
+	d.client = resourceData.client
+}
+
+// Metadata returns the data source type name.
+func (d *LoggingEndpointsDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_mte_logging_endpoints"
+}
+
+// Schema defines the schema for the data source.
+func (d *LoggingEndpointsDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"endpoints": schema.ListNestedAttribute{
+				Computed: true,
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"id": schema.StringAttribute{
+							Computed: true,
+						},
+						"type": schema.StringAttribute{
+							Computed: true,
+						},
+						"environmentid": schema.StringAttribute{
+							Computed: true,
+						},
+						"config": schema.SingleNestedAttribute{
+							Computed: true,
+							Attributes: map[string]schema.Attribute{
+								"dataset": schema.StringAttribute{
+									Computed: true,
+								},
+								"projectid": schema.StringAttribute{
+									Computed: true,
+								},
+								"table": schema.StringAttribute{
+									Computed: true,
+								},
+								"email": schema.StringAttribute{
+									Computed: true,
+								},
+								"headers": schema.ListNestedAttribute{
+									Computed: true,
+									NestedObject: schema.NestedAttributeObject{
+										Attributes: map[string]schema.Attribute{
+											"columnname": schema.StringAttribute{
+												Computed: true,
+											},
+											"headername": schema.StringAttribute{
+												Computed: true,
+											},
+											"defaultvalue": schema.StringAttribute{
+												Computed: true,
+											},
+										},
+									},
+								},
+								"secretkey": schema.StringAttribute{
+									Computed: true,
+									Optional: true,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// Read refreshes the terraform state with the latest data.
+func (d *LoggingEndpointsDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var state LoggingEndpointsDataSourceModel
+
+	loggingEndpoints, err := d.client.ReadMTELoggingEndpoints()
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to get logging endpoints from Altitude provider",
+			err.Error(),
+		)
+		return
+	}
+
+	for _, endpoint := range loggingEndpoints.Endpoints {
+		endpointState := LoggingEndpointDataSourceModel{
+			Type:          types.StringValue(endpoint.Type),
+			EnvironmentId: types.StringValue(endpoint.EnvironmentId),
+			Config:        transformToConfigResourceModel(endpoint.Config),
+			ID:            types.StringValue("placeholder"),
+		}
+
+		state.Endpoints = append(state.Endpoints, endpointState)
```

Comment:
Can this use the same format as `headerModels[i] = headersBody` with the `i` for consistency. Or the one below use append. ([i] might be quicker than append though, not sure)

Reply:
done

Reply:
I disagree as the headerModels[i] = headersBody is replacing items in the list, the list already exists so append wouldn't apply here. Whereas for state.Endpoints, the list doesnt exist. So state.Endpoints[i] doesnt exist and an error will be thrown.